File: 18000/100/BJ_18185_라면_사기_Small/BJ_18185_라면_사기_Small.java
// https://www.acmicpc.net/problem/18185

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;

public class BJ_18185_라면_사기_Small {
    public static void main(final String[] args) throws IOException {
        try (final var stdin = new BufferedReader(new InputStreamReader(System.in))) {
            final var N = Integer.parseInt(stdin.readLine());
            final var noddles = new int[N];
            {
                final var tokens = stdin.readLine().split(" ");
                for (var i = 0; i < N; ++i) {
                    noddles[i] = Integer.parseInt(tokens[i]);
                }
            }

            final var bundles = new ArrayDeque<Integer>();
            var numBundleRemains = 0;
            var cost = 0;
            for (var i = 0; i < N; ++i) {
                final var numBundleOrders = Math.min(numBundleRemains, noddles[i]);
                final var numNewOrders = Math.max(noddles[i] - numBundleOrders, 0);
                cost += numNewOrders * 3 + numBundleOrders * 2;

                final var numNewBundleOrders = Math.min((i < N - 1 ? noddles[i + 1] : 0), numNewOrders);
                bundles.addLast(numNewBundleOrders);

                numBundleRemains += numNewBundleOrders - (2 <= i ? bundles.pollFirst() : 0);
            }

            System.out.print(cost + "\n");
        }
    }
}

File: 18000/100/18185_라면_사기_Small/BJ_18185_라면_사기_Small.py
# https://www.acmicpc.net/problem/18185

import sys


def buy(facs, lb, ub, n):
    for i in range(lb, ub):
        facs[i] -= n
    return [3, 5, 7][ub - lb - 1] * n


readline = lambda: sys.stdin.readline().strip()
N = int(readline())
facs = [*map(int, readline().split()), 0, 0]

cost = 0
for i in range(0, N):
    if 0 < (n := min(facs[i:i + 3])):
        if facs[i] < facs[i + 1] and facs[i + 2] < facs[i + 1]:  # 요철 모양 처리
            cost += buy(facs, i, i + 2, min(facs[i], facs[i + 1] - facs[i + 2]))

    for sz_window in range(3, 0, -1):
        cost += buy(facs, i, i + sz_window, n) if 0 < (n := min(facs[i:i + sz_window])) else 0

print(cost)
